### ISSUE TITLE
BIG: Breaking everything into smaller repos

### ISSUE THREAD

**Issue by onkoe** [Sooner-Rover-Team/SoonerRoverTeamV#32]
### What are we doing? 
Currently, the Rover Team's repository is really... scary. Frankly, I'm not too sure how everyone got so much work done with this level of organization. Anyways, the plan is to fix the problems by organizing all of the different subdirectories that are sitting in this one repo. As needed, they will all become separate GitHub repos on the Sooner Rover Team org. 

### How are we splitting everything up? 
I'm deciding a rough draft below. You can use the command `tree` to easily look at the file structures. Use `tree -d` to see only directories (folders). When I see patterns of grouped utilities that are developed together, they'll stay as a group. Otherwise, they're getting stored or archived. 

### How will it look? 
Each of the listed items below will become their own split-off repo and project. The `(/foo)` represents its current location in the giant repo. There don't need to be any active maintainers, since our "human resources" are limited. That said, if something needs to be archived, let me know.  Also, this link on [moving files between repos](https://medium.com/@ayushya/move-directory-from-one-repository-to-another-preserving-git-history-d210fa049d4b) while preserving commit history will be *very* useful!

Anyways, I plan to have it look something like this: 

##### (/Arduino/)
This was a pretty easy split. This folder contains various Arduino files, all of which are related. Here's a folder listing: 
```
Arduino/
├── arm_controller
├── arm_test
├── capstone_arm
│   ├── control_arm
│   ├── rover_arm
│   ├── rover_arm_pid_test
│   └── rover_arm_test
├── drill_controller
├── drill_test
├── servo_test
└── wheel_controller
```

##### (/Autonomous/YOLO/darknet/)
This is a sizeable library whose source is just sitting in the middle of everything. It really needs its own repo. Here's a folder listing: 
```
Autonomous/
├── YOLO
│   └── darknet
│       ├── 3rdparty
│       │   ├── pthreads
│       │   │   └── include
│       │   └── stb
│       │       └── include
│       ├── cfg
│       ├── cmake
│       │   └── Modules
│       ├── data
│       │   └── soro
│       ├── include
│       ├── scripts
│       │   ├── log_parser
│       │   └── windows
│       └── src
│           └── csharp
├── artags
│   └── old
│       └── PrintARTags
├── examples
├── gps
└── libs
```

##### (/MBED/)
MBED is a mixed bag for me. It includes many small projects that have connected goals. For this reason, I'm not sure if I want to put each project in its own repo, do that but with submodules, or just keep everything together. For now, I'll just say it belongs together. Here's a listing (it's so small that you can put files too): 
```
MBED
├── EBoxMBED
│   └── main.cpp
├── README.md
├── SPItoCANRead
│   └── main.cpp
├── SPItoCANSend
│   └── main.cpp
├── UDPSocket
│   └── main.cpp
└── WheelMBED
    └── main.cpp
``` 

##### (/Mission Control/RoverMap/)
To be honest, even after studying the README, I'm not sure if these are two separate projects (frontend and backend) or if they both require each other. For now, let's keep them together. Here's a directory listing: 
```
RoverMap/
├── RoverMapTileGenerator
├── example
└── frontend
    └── leaflet
        └── images
```
(frontend is a whole web interface. but it's a lil guy)

##### (/Mission Control/python_drive/)
This project is completely distinct from the rest of everything. It controls the vehicle! Here's a short file listing: 
```
python_drive/
├── README.md
├── config.ini
├── controllertester.py
├── manual_operate.py
└── util.py
```

##### (/Mission Control/video_streamer/)
The video_streamer project has a lot of documentation! I think it would be well off in its own repo, along with a detailed INSTALLATION.md and reduced README.md. Here's a file listing: 
```
video_streamer/
├── README.md
├── client.py
├── config.ini
├── static
│   └── logo.png
├── templates
│   └── index.html
├── utils.py
└── videostream.py
```

### TO ARCHIVE
- udp_mc_drill and udp_mc_drive **(/Mission Control/udp_mc_dr\*\*\*)**
    - According to the Mission Control README, both of these are useless and have been replaced with a modern Python script. 
    - It's sad to get rid of the only Rust projects, but necessary :(

### And..?
With that, we're done! All of the files in the Sooner Rover Team's modern repo have been accounted for! I think it would be also wise to make a **Junkyard** repo that acts as a place to store archived projects that are pretty old/small. If you have any suggestions or changes I should make, please let me know!

### Complete

- [ ] Arduino
- [ ] darknet
- [ ] MBED
- [ ] RoverMap
- [ ] python_drive
- [ ] video_streamer
- [ ] Graveyard
    - [ ] udp_mc_drill
    - [ ] udp_mc_drive
    - [x] my rusty dreams :(

**Comment by bentonsmith8** [comment]
I like the breakdown for the most part. A few things can be moved around even more such as capstone_arm which hasn't been used in years and can go into the graveyard. Also we'll still need a repo for the autonomous code itself (stuff in gps, libs, and examples).

**Comment by cwealeft** [comment]
The arm capstone is current though… Christine WealeSenior, El.Eng. Junior, OUVP of Service, Alpha Phi OmegaUG Research Asst. | RIL From: Benton SmithSent: Friday, January 27, 2023 12:47To: Sooner-Rover-Team/SoonerRoverTeamVCc: SubscribedSubject: Re: [Sooner-Rover-Team/SoonerRoverTeamV] BIG: Breaking everything into smaller repos (Issue #32) I like the breakdown for the most part. A few things can be moved around even more such as capstone_arm which hasn't been used in years and can go into the graveyard. Also we'll still need a repo for the autonomous code itself (stuff in gps, libs, and examples).—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

**Comment by bentonsmith8** [comment]
As far as I can tell, all of the code in that folder is from the OLD arm capstone from 4-5 years ago. I'm not sure if any of it applies since that one worked very differently to the arm that we currently use on the rover.

**Comment by onkoe** [comment]
Is graveyarding this the consensus? I'm happy to start working on moving everything, and of course, we can always revert and move the *capstone_arm* project somewhere else if it's deemed necessary.

**Comment by bentonsmith8** [comment]
Yeah, I like the idea of just throwing everything obsolete in a repo by itself so it's not in the way of the useful stuff but we have access to it if we need it.

**Comment by ROMANT21** [comment]
For the Autonomous Folder, I would say that you could move the YOLO folder into a sub-repo. Other than that, I give it a green light.

**Comment by onkoe** [comment]
dont touch til march 4th :)

**Comment by onkoe** [comment]
Note: `MBED` was dissolved into `Microcontrollers` with fd11b43. `Arduino` was also moved. 

In the future, please explicitly commit for moves like this - it's hard to keep track! Squash merges are a little scary ;)

**Comment by onkoe** [comment]
Also moved: 
- [x] Discord Bot
- [x] PCB
    - This seems to be a CAD backup repo. GitHub doesn't like that, and `git` isn't designed for it. It's best to remove this repo ASAP! 

I now need to apply the MIT license to each of the moved repos. Let's keep track:
- [ ] pcb
- [ ] discord_bot
- [ ] Autonomous
- [ ] gps
- [ ] rovercam-python
- [ ] python_drive
- [ ] RoverMap
- [ ] YOLO (wish we'd have kept it WTFPL ;) )
- [ ] Microcontrollers
- [ ] soro-documentation
